odoo/addons/base/models/ir_rule.py

```diff
@@ -143,6 +184,26 @@ def write(self, vals):
         self.clear_caches()
         return res
 
+    def _make_access_error(self, operation, records):
+        _logger.info('Access Denied by record rules for operation: %s on record ids: %r, uid: %s, model: %s', operation, records.ids, self._uid, records._name)
+
+        if not self.env.user.has_group('base.group_no_one'):
+            return AccessError(_('The requested operation cannot be completed due to security restrictions. Please contact your system administrator.\n\n(Document type: %s (%s), Operation: %s)') % \
+                            (records._description, records._name, operation))
+
+        # debug mode, provide more info
+        rules = self._get_failing(records, mode=operation).sudo()
+        return AccessError(_("""The requested operation ("%s" on "%s") was rejected because of the following rules:
+%s
+%s
+(records: %s, uid: %d, model: %s)""") % (
+            operation, records._description,
+            '\n'.join('- %s' % rule.name for rule in rules),
+            _('\nNote: this might be a multi-company issue\n') if any(
+                'company_id' in r.domain_force for r in rules) else '',
+            list(records.ids), self.env.user.id, records._name
```

Comment:
As above, you might want to limit the number of IDs being displayed in the message, to avoid killing the UI in case of an error with millions of records.